lib/spack/spack/solver/asp.py

```diff
@@ -970,6 +977,13 @@ def _model_has_cycles(self, models):
 
         return cycle_result.unsatisfiable
 
+    def _compiler_flag_has_propagation(self, flags):
+        for flag in flags:
+            for flag_type, flag_vals in flag.items():
+                if any(val.propagate for val in flag_vals):
+                    return True
+        return False
```

Comment:
This method is hard to read. Could you try replacing it with:

```suggestion
        return any(val.propagate for _, flag_vals in flags.items() for val in flag_vals)
```
Or the equivalent non-comprehension form
```
        for _, flag_vals in flags.items()
            if any(val.propagate for val in flag_vals):
                return True
        return False
```
And see the comment on the calling code for the equivalent change there.

I think the nested comprehension is probably easier to maintain, but I'm not sure of that by any means.